### DISCUSSION TITLE
Need to add a way for users to be able to see rank specific information on forms

### DISCUSSION THREAD

**Discussion by tlammer** [specify/specify7#5537] [Feature Requests]
Edinburgh has mentioned they need to have a way to see the family # and genus # for their determinations, ideally on the determination form.  The issue is there is no easy way to do that because there may be ranks in between family and genus or genus and species, so even in a query, you couldn't reliably get back the number for all objects because it goes based on parent or parent of parent, rather than rank, so if the parent is subgenus rather than genus, you don't get the number.  And currently, there is no way to add a field that shows taxon information for ranks other than the lowest determined node.

**Comment by norinespears** [comment]
This is not specific to Sp 7
This needs a longer discussion

**Comment by maxpatiiuk** [comment]
This should be possible though a formatter (by creating a recursive formatter).
Would be made even easier by `xml-editor` branch.

Example:
Let's say you want to see family author. Create a formatter where
If `treedefitem.rankid=<family rank id>`, then display autor
Else If `treedefitem.rankid=<root rank id>`, then display empty field (or some sort of fallback) - this is in case the taxon related to the determination is higher than family
Else, display parent node, using this very same formatter

**Comment by chanulee1** [comment]
How do I test this?

**Comment by maxpatiiuk** [comment]
Here is an example of how recursive formatter may look like:

![image](https://user-images.githubusercontent.com/40512816/232093308-544ac991-1bc7-4339-938b-25ab132193f5.png)

This says if taxon rank name is "Order", then display "Source" field, otherwise, format the parent of the taxon using the same formatter (thus in effect climbing the tree). If it happens that the current node is above Order, it will just climb the tree until it reaches root node, at which point the parent would be null so formatter would produce an empty string.

Also you don't have to worry about accidentally creating infinite recursion, as the formatter code has recursion detection.

One thing to note though: the live preview feature does not work for recursive formatters if you made changes (it will continue to show the formatter before you made the changes). To fix this, create your recursive formatter, save it, reload the page, then open it again, and then the formatter will work like expected. Opened a bug for that - https://github.com/specify/specify7/issues/3347

**Comment by grantfitzsimmons** [comment]
It may be advisable to change "Condition" to "Condition Field Value" to clarify how the system works

**Comment by chanulee1** [comment]
Looks good. Recursive formatters appear to be working and allow displaying rank specific information.

**Comment by grantfitzsimmons** [comment]
You can't define a format explicitly on the form. Not closed...

**Comment by grantfitzsimmons** [comment]
See this discussion Discourse to see another request:
https://discourse.specifysoftware.org/t/show-family-in-the-determination-form/1902/4?u=grant
